Add warning package

diegohaz opened the pull request:
## Description

The need of a `warning` utility was brought to light by @gziolo in several issues:

- https://github.com/WordPress/gutenberg/pull/18931#discussion_r359833402
- https://github.com/WordPress/gutenberg/issues/12535#issuecomment-549249354
- https://github.com/WordPress/gutenberg/issues/11631#issuecomment-544858908

## How has this been tested?

[Unit tests](https://github.com/WordPress/gutenberg/blob/6649664f5595a5196214e604066f6e65fe0e7191/packages/warning/src/test/index.js)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

diegohaz committed:
Add new warning package

diegohaz committed:
Add babel-plugin-dev-expression

gziolo commented on babel.config.js:
```diff
@@ -6,6 +6,7 @@ module.exports = function( api ) {
 		plugins: [
 			'babel-plugin-emotion',
 			'babel-plugin-inline-json-import',
+			'babel-plugin-dev-expression',
```
Ideally, there would be an option in the plugin which ensures that a random function call with the caller name warning doesn’t get wrapped with the guard condition. In this case tracking the important statement would do the trick.

diegohaz replied:
I removed `babel-plugin-dev-expression` and included a `@wordpress/warning/babel-plugin` file in the package that checks the specifier name of any import declaration pointing to `@wordpress/warning`. It'll only transform function calls with that name.

gziolo reviewed (commented):
This is great! Thank you for starting this PR 👍

There is this question whether we want to have this Babel plugin included or we create our own variation enabled in the preset we expose for 3rd party projects for developers convenience.

There is one more thing to consider, how to wire the package into WordPress globals. It uses the default export so on principle it should be whitelisted in Webpack config:
https://github.com/WordPress/gutenberg/blob/fb836a403f805be364eb42bbde810c22ef29e021/webpack.config.js#L85-L93

Well, it’s a quite interesting case as we probably don’t want to promote it’s usage in production mode. It probably doesn’t harm to do it for consistency. What it does it changes how this default export gets exposed as global from wp.warning.default to wp.warning.

diegohaz committed:
Remove babel-plugin-dev-expression

diegohaz committed:
Add babel-plugin to the warning pacakge

diegohaz committed:
Include @wordpress/warning/babel-plugin into @wordpress/babel-preset-default

diegohaz committed:
Add warning to webpack.config.js

diegohaz commented:
Thanks for the review @gziolo! I've replaced `babel-plugin-dev-expression` with a custom version that ignores random `warning` calls that aren't pointing to `@wordpress/warning`. This is exposed as `@wordpress/warning/babel-plugin`.

I've also included that in `@wordpress/babel-preset-default` and added the entry on `webpack.config.js`.

This is the first time I wrote a babel plugin. Any suggestion on how to write it better is welcome! :)

aduth commented on packages/warning/src/babel-plugin.js:
```diff
@@ -0,0 +1,62 @@
+					// Turns this code:
+					// warning(condition, argument, argument);
+					// into this:
+					// process.env.NODE_ENV !== "production" ? warning(condition, argument, argument) : void 0;
```
I'm very glad you are thinking about dead code elimination with this pull request.

One concern I have about this specific implementation is that it makes a few assumptions about the runtime and build environments that I'm not sure we should be comfortable to make:

1. Something else in the build will ensure that `process.env.NODE_ENV` global is assigned, or that the generated code is only able to run in Node.js (and not in a browser, where `process.env` does not exist)
2. Something else in the build will perform dead code elimination on unreachable code (like Webpack + Uglify/Terser).

If either of these are not met, then either the code will not in-fact be removed, or at worst it can cause errors in the runtime environment (e.g. trying to access property `NODE_ENV` of an undefined `process.env` in a browser).

What I'd wonder is: Is there any reason we'd not want to just remove the call expression here and now based on what `process.env` evaluates to in the Babel transform? I guess I can see how we might not want that though if, for example, we want the warnings to be logged in projects which depend on our packages (where the published packages would be built in a "production" environment?).

It all looks very similar to Facebook's internal `invariant` utility. I guess this is intentional? How do they handle it? Or, since those are used in internal projects, are they fine to make those assumptions about the runtime/build environment?

Alternatively, we could do some combination of:

- Document the build environment assumptions
- Check that `process` global exists before evaluating it
  - e.g. `typeof process !== 'undefined' && typeof process.env !== 'undefined' && process.env.NODE_ENV`

diegohaz replied:
@aduth I think I assumed we were handling `process.env.NODE_ENV` somehow because it's being used in other parts of the code base. For example: https://github.com/WordPress/gutenberg/blob/8719578ea93cc2c51d843627d5a50910cf6c65d7/packages/components/src/tooltip/index.js#L147-L150

Looking at the code generated in the `build`/`build-module` folders, that check remains there. I'm confused now! How does it work? Where can I find the code responsible for building the code for the browser?

> What I'd wonder is: Is there any reason we'd not want to just remove the call expression here and now based on what process.env evaluates to in the Babel transform? I guess I can see how we might not want that though if, for example, we want the warnings to be logged in projects which depend on our packages (where the published packages would be built in a "production" environment?).

I expect some warnings like this:

```js
warning(
  dialog.tabIndex === undefined || dialog.tabIndex < 0,
  'It\'s recommended to have at least one tabbable element inside dialog. The dialog element has been automatically focused.',
  'If this is the intended behavior, pass `tabIndex={0}` to the dialog element to disable this warning.'
);
```

If we remove the warning call from the code we ship to npm, then people won't see that (when using `@wordpress/components`, for example).

gziolo replied:
Facebook uses `__DEV__`:

https://github.com/facebook/react/blob/3c54df0914f02fed146faa519a5a899d0e3af32e/packages/shared/consoleWithStackDev.js#L31

I guess the benefit is that it fallbacks to `undefined` and therefore doesn't call the guarded code. You can still benefit from dead code elimination by teaching webpack to convert it to `true` or `false`.

I don't know what's the best way to move forward, but I'm sharing how I see the approach FB took.

aduth replied:
> I guess the benefit is that it fallbacks to `undefined` and therefore doesn't call the guarded code. You can still benefit from dead code elimination by teaching webpack to convert it to `true` or `false`.

In this case, I assume something must still be assigning `__DEV__` value, as otherwise the reference to an undeclared variable would throw an error:

```js
( function() {
    if ( __DEV__ ) {
        console.log( 'Constant assigned' );
    } else {
        console.log( 'Constant not assigned' );
    }
}() )
// Uncaught ReferenceError: __DEV__ is not defined
```

Revisiting my earlier comment, I'm actually not sure we're doing anything wrong here with how we reference `process.env.NODE_ENV`. I think it could be fair to say that if we're distributing this as a Node package, it may be more-or-less assumed that it be evaluated as in a Node context where the `process` global is defined. I do worry that in a standalone context, there is a chance that this might not be substituted correctly when building for a browser target. At least in the case of Webpack, this is handled automatically as far as I can tell (see [`mode`](https://webpack.js.org/configuration/mode/), [`DefinePlugin`](https://webpack.js.org/plugins/define-plugin/)).

A safe option might just be to test its presence before evaluating it (`typeof process !== 'undefined'`).

With regards to other existing code in the codebase, I think one of two of the following might be the case:

- We're okay to assume the code is not used outside of Gutenberg, or at least in a context not already subjected to Gutenberg's own build process
- Or, that code is prone to issues as well

For something like a `warning` package, it seemed highly likely we'd want to promote that this be used separate from Gutenberg and its build, hence the original concern.

diegohaz committed:
Merge branch 'master' into add/warning

diegohaz committed:
Merge branch 'master' into add/warning

diegohaz committed:
Check the presence of process.env

diegohaz commented:
Added a check on the presence of `process.env` (be1f9213ddb3cf4a2fbca0ca20e08e81edad3ce4)

gziolo commented:
> Added a check on the presence of `process.env` ([be1f921](https://github.com/WordPress/gutenberg/commit/be1f9213ddb3cf4a2fbca0ca20e08e81edad3ce4))

As far as I can tell, in the version published to npm, we still will see `process.env.NODE_ENV` check added from the Babel plugin.

> For something like a `warning` package, it seemed highly likely we'd want to promote that this be used separate from Gutenberg and its build, hence the original concern.

Still, this should be fine with this proposal. If you don't use the Babel plugin, it should work in all cases.

aduth commented on packages/babel-preset-default/index.js:
```diff
@@ -56,6 +56,7 @@ module.exports = function( api ) {
 		presets: [ getPresetEnv() ],
 		plugins: [
 			require.resolve( '@babel/plugin-proposal-object-rest-spread' ),
+			require.resolve( '@wordpress/warning/babel-plugin' ),
```
_(**Edit:** I noticed before submitting review that there _is_ a `babel-plugin.js` in the top-level directory. I must have missed it in my first pass. All the same, worth considering the notes about transpilation, especially considering that the top-level directory which would be imported here is referencing _the source_ and not the built file, which means it's pointless for us to include in `src`, and also the built file in `build` is unused)_

I'm a bit confused about how this path actually gets resolved, considering that `babel-plugin` is contained within a `src` directory (or one of the `build(-)` directories after build). For referencing it by this path, I might have expected that it would require that a `babel-plugin.js` file exist at the top-level directory. Is that not the case?

It's also worth pointing out that, unlike other packages where we need to transpile the code using Babel (to be able to run in a browser or other environment with non-modern JS support), I expect that the code of the Babel plugin could be run directly, and not require to be part of the build. There are examples of prior art for this elsewhere, e.g. [`@wordpress/babel-plugin-import-jsx-pragma`](https://github.com/WordPress/gutenberg/tree/master/packages/babel-plugin-import-jsx-pragma), which does not have `src`, and as such is not [built](https://github.com/WordPress/gutenberg/blob/5714888c0acd52fedfc31a64815c25e400f6caac/bin/packages/build.js#L125).

This one is a bit unique in that we have both of: code which doesn't need to be transpiled, and which _does_ need to be transpiled (the `warning` function itself). I don't think there's much harm in just transpiling it all, but I wanted to point it out all the same. And it might be relevant for if we do need a `babel-plugin.js` at the top-level directory (since we could just move it verbatim from where it currently resides).

diegohaz replied:
You're right! I moved the `babel-plugin.js` file to the package top-level directory together with its `test` folder, to be consistent with the other babel plugin packages in the repo.

aduth commented on package-lock.json:
```diff
@@ -22850,7 +22854,7 @@
 				"clone-deep": {
 					"version": "0.2.4",
-					"resolved": "http://registry.npmjs.org/clone-deep/-/clone-deep-0.2.4.tgz",
+					"resolved": "https://registry.npmjs.org/clone-deep/-/clone-deep-0.2.4.tgz",
```
This value has a bad tendency to flip back and forth in local environments. It's not strictly part of your changes here, but I think I've seen the same locally, so it might be good to include this anyways, just to avoid future issues.

gziolo replied:
Rights, it always shows up for me locally as https. I even updated it in `master` a few days ago:
https://github.com/WordPress/gutenberg/commit/fe258a64eed124597e37c5f0ed588e395241bf25

There might be something going on with operating systems or versions of npm client.

aduth reviewed (commented):
@gziolo It's unclear to me from your comment: Are you expecting any other changes here?

It's generally looking quite good to me. My previous point still stands that we're expecting that someone has dead-code elimination as part of their build pipeline for removing the string literals. But we're not documenting this anywhere. In fact, we're documenting that the custom Babel plugin alone would be enough to eliminate the string literals. It might be something where it's enough to include some extra documentation that Uglify or Terser should be used in combination with the Babel plugin (and `process.env.NODE_ENV` substitution) in order for the strings to be removed from a production bundle.

diegohaz committed:
Check presence of process.env in the code transformed by the babel plugin

diegohaz committed:
Move babel-plugin.js and its test file to the directory top-level

diegohaz committed:
Improve README with information about dead code removal

diegohaz committed:
Turn sideEffects into true in package.json

diegohaz commented:
> As far as I can tell, in the version published to npm, we still will see process.env.NODE_ENV check added from the Babel plugin.

Updated the babel plugin to check the presence of `process.env` in the transformed code as well (0765740a0701f1bdab1be9a307a360ca0937bb55).

> My previous point still stands that we're expecting that someone has dead-code elimination as part of their build pipeline for removing the string literals. But we're not documenting this anywhere. In fact, we're documenting that the custom Babel plugin alone would be enough to eliminate the string literals. It might be something where it's enough to include some extra documentation that Uglify or Terser should be used in combination with the Babel plugin (and process.env.NODE_ENV substitution) in order for the strings to be removed from a production bundle.

Updated the README with more information about dead code elimination (b045842ba269f0539f3327e947e40f521fa86207). It can be improved though.

diegohaz committed:
Revert sideEffects change

diegohaz committed:
npm run docs:build

gziolo commented:
I tested it with the following changes:

```diff
diff --git a/package-lock.json b/package-lock.json
index 3c8c07c44..ce01de8a8 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -8251,7 +8251,8 @@
 			"version": "file:packages/a11y",
 			"requires": {
 				"@babel/runtime": "^7.4.4",
-				"@wordpress/dom-ready": "file:packages/dom-ready"
+				"@wordpress/dom-ready": "file:packages/dom-ready",
+				"@wordpress/warning": "file:packages/warning"
 			}
 		},
 		"@wordpress/annotations": {
diff --git a/packages/a11y/package.json b/packages/a11y/package.json
index eb2f11f3a..a28f44ad5 100644
--- a/packages/a11y/package.json
+++ b/packages/a11y/package.json
@@ -23,7 +23,8 @@
 	"react-native": "src/index",
 	"dependencies": {
 		"@babel/runtime": "^7.4.4",
-		"@wordpress/dom-ready": "file:../dom-ready"
+		"@wordpress/dom-ready": "file:../dom-ready",
+		"@wordpress/warning": "file:../warning"
 	},
 	"publishConfig": {
 		"access": "public"
diff --git a/packages/a11y/src/index.js b/packages/a11y/src/index.js
index cb79e767b..d7f0f76e5 100644
--- a/packages/a11y/src/index.js
+++ b/packages/a11y/src/index.js
@@ -2,6 +2,7 @@
  * WordPress dependencies
  */
 import domReady from '@wordpress/dom-ready';
+import warning from '@wordpress/warning';
 
 /**
  * Internal dependencies
@@ -14,6 +15,7 @@ import filterMessage from './filterMessage';
  * Create the live regions.
  */
 export const setup = function() {
+	warning( true, 'How is it going?' );
 	const containerPolite = document.getElementById( 'a11y-speak-polite' );
 	const containerAssertive = document.getElementById( 'a11y-speak-assertive' );
 ```

I can confirm that the warning is printed on the JS console only for the development build and I can't find any referenced to the warning package in the production mode.

gziolo reviewed (approved):
> @gziolo It's unclear to me from your comment: Are you expecting any other changes here?
>
> It's generally looking quite good to me. My previous point still stands that we're expecting that someone has dead-code elimination as part of their build pipeline for removing the string literals. But we're not documenting this anywhere. In fact, we're documenting that the custom Babel plugin alone would be enough to eliminate the string literals. It might be something where it's enough to include some extra documentation that Uglify or Terser should be used in combination with the Babel plugin (and `process.env.NODE_ENV` substitution) in order for the strings to be removed from a production bundle.

I wasn't really promoting any further code changes but I see that @diegohaz applied them regardless. In addition, the proposed solution works out of the box in all cases so now it feels that we could leave it as is if we don't care as much about having the `process` polyfill included in the development mode only.

@aduth - I'd defer the final call to you, from my perspective this patch is good to go. I'm fine with leaving it as is and it's also fine to remove the check from the Babel plugin in favor of the comment in the package description.

diegohaz committed:
Update packages/warning/babel-plugin.js

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

aduth commented on packages/warning/babel-plugin.js:
```diff
@@ -0,0 +1,79 @@
+				// Ignore if it's already been processed
+				if ( node[ seen ] ) {
+					return;
+				}
```
I'm a bit confused on this. Why would this ever happen?

diegohaz replied:
I don't fully understand how babel does the transformation. I initially wrote the plugin without this check, but then I get this error when running babel:
```
RangeError: Maximum call stack size exceeded
```
You can reproduce it by commenting that snippet and running `npm run test-unit packages/warning`.

I guess it tries to recursively transform the code. If that's the case, it'll always find the same `warning(...)` call wrapped within the `process.env.NODE_ENV` check.

aduth replied:
> I guess it tries to recursively transform the code. If that's the case, it'll always find the same `warning(...)` call wrapped within the `process.env.NODE_ENV` check.

Ah, that makes total sense to me in retrospect. Thanks for clarifying.

It would be nice if we could somehow inform the transform traversal to skip that node at the time we're creating it, but it's not immediately clear to me how we'd do that without transforming it in some other way such that it's not picked up by the normal procedure.

I could imagine something in the future where we could tie this into some "opt out" mechanism. An escape hatch something much like [ESLint's `eslint-disable-line`](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments). Where the procedure for replacing the node would mark the replacement as skipped.

I think this is fine for now, though.

aduth commented on packages/warning/src/index.js:
````diff
@@ -0,0 +1,45 @@
+ * ```
+ */
+export default function warning( condition, ...messages ) {
+	if ( isDev() ) {
````
Minor: You could avoid indentation on the entire function with an early return (similar to what's done on the immediate next line with `if ( ! condition ) {`):

```js
if ( ! isDev() ) {
	return;
}
```

Alternatively, you could assign the function itself as a noop for the static export member, rather than evaluating each time it's called (not that it's a very expensive evaluation):

```js
const warning = isDev() ?
	() => {} :
	( condition, ...messages ) => {
		// ...
	};

export default warning;
```

aduth commented on packages/warning/src/index.js:
```diff
@@ -0,0 +1,45 @@
+			return;
+		}
+
+		const text = messages.join( '\n' );
```
Not really concerned too much over this, but was there any prior art for choosing `\n` here?

Notably, if we had implemented this as something like:

```js
console.warn( ...messages );
```

`warn` would join _using a space_.

Technically speaking, for `console` logging, this behavior is not specified, and left up to the implementation, so a newline could be reasonable as well:

>How the implementation prints args is up to the implementation, but implementations should separate the objects by a space or something similar, as that has become a developer expectation.

https://console.spec.whatwg.org/#printer

aduth replied:
This also raises a question: Why do we support multiple `message` strings? Is there a demonstrated need for it? Would it just be simpler to accept only a single string, and consider to add support for the variadic arguments if it becomes a problem? Especially considering the specific treatment of `messages` is not currently documented very thoroughly, it seems like a feature that [we might not need](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

diegohaz replied:
I tried to mimic the same API as the `warning` function on Reakit, [which uses the multiline feature a lot](https://github.com/reakit/reakit/search?q=warning&unscoped_q=warning). But I completely agree that it's better to start with a single parameter for now.

aduth replied:
Thanks for linking to the prior art. That definitely helps clarify the way in which you were expecting it to be used. At least for some of the examples I see there, it reminds me a bit of this project's [`@wordpress/deprecated` module](https://github.com/WordPress/gutenberg/tree/master/packages/deprecated), at least so far as having some consistency in how those messages are structured (a "title" line, a "see more" line, etc). The API for `deprecated` is not always obvious, and I can see merit in having the low-level APIs like `warning` be unopinionated about this sort of thing. But I do think it's something in the future where if we start to see these sorts of patterns emerging, we can decide then whether it's something we want to leave non-formalized (and optimize for multi-line messages) or build other abstractions around (something more like `deprecated` which manages how the message is structured in a consistent way).

aduth committed:
warning: Add types checking for warning package

aduth reviewed (approved):
This is looking good to me.

The one thing I would propose to consider is the review comment concerning whether we want to make the variadic `messages` part of the public API.

I also pushed a commit 44cdafec2 to add this package to TypeScript types checking, to avoid follow-on work related to #18838. I hope you don't mind. This also includes type hints for the Babel plugin, which helped to verify there were no issues with the implementation there.

diegohaz committed:
Merge branch 'master' into add/warning

diegohaz committed:
Avoid indentation on the warning function with an early return

diegohaz committed:
Fix TS error on babel-plugin.js

diegohaz committed:
Accept a single string as message instead of multiple messages
